### ISSUE TITLE
Low iperf3 throughput with sonata-fw-vnf-docker image

### ISSUE THREAD

**Issue by knodir** [sonata-nfv/son-examples#90]
Hi,

Can you please take a look at my [example](https://gist.github.com/knodir/646af08c5c63394d64dce52575765681) and help me identify the network bottleneck at `sonata-fw-vnf-docker` image? I created `client - firewall - server` chain and trying to measure performance of the chain with iperf3. The client can ping server, but iperf3 reports 0 bits/sec on the server as follows:

```
root@client:/# iperf3 -c 10.0.0.5 -t 10 -b 100M                                                                                                               
Connecting to host 10.0.0.5, port 5201
[  4] local 10.0.0.2 port 44502 connected to 10.0.0.5 port 5201
[ ID] Interval           Transfer     Bandwidth       Retr  Cwnd
[  4]   0.00-1.00   sec  87.7 KBytes   718 Kbits/sec   15   1.41 KBytes       
[  4]   1.00-2.00   sec  0.00 Bytes  0.00 bits/sec    1   1.41 KBytes       
[  4]   2.00-3.00   sec  0.00 Bytes  0.00 bits/sec    0   1.41 KBytes       
[  4]   3.00-4.00   sec  0.00 Bytes  0.00 bits/sec    1   1.41 KBytes       
[  4]   4.00-5.00   sec  0.00 Bytes  0.00 bits/sec    0   1.41 KBytes       
[  4]   5.00-6.00   sec  0.00 Bytes  0.00 bits/sec    0   1.41 KBytes       
[  4]   6.00-7.00   sec  0.00 Bytes  0.00 bits/sec    1   1.41 KBytes       
[  4]   7.00-8.00   sec  0.00 Bytes  0.00 bits/sec    0   1.41 KBytes       
[  4]   8.00-9.00   sec  0.00 Bytes  0.00 bits/sec    0   1.41 KBytes       
[  4]   9.00-10.00  sec  0.00 Bytes  0.00 bits/sec    0   1.41 KBytes       
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bandwidth       Retr
[  4]   0.00-10.00  sec  87.7 KBytes  71.8 Kbits/sec   18             sender
[  4]   0.00-10.00  sec  0.00 Bytes  0.00 bits/sec                  receiver
```

I get around 28 Gbps when I replace the firewall image with [sonata-snort-ovs-vnf-docker](https://github.com/sonata-nfv/son-examples/tree/master/vnfs/sonata-snort-ovs-vnf-docker) in `client - snort - server` setup.

Note that `knodir/sonata-fw-vnf` image I use in the example is the same as [sonata-fw-vnf-docker](https://github.com/sonata-nfv/son-examples/tree/master/vnfs/sonata-fw-vnf-docker) but also has tcpdump for debugging.

I feel there is some problem with `ovs1` bridge and its ports (see firewall image [Dockerfile](https://github.com/sonata-nfv/son-examples/blob/master/vnfs/sonata-fw-vnf-docker/start_fw.sh#L28)), but I could not unearth it after 6 hours of debugging. So, any help is appreciated. 

Update: I forgot to mention the main observation. Establishing iperf3 connection takes a long time (~2-3mins) and once the connection is established, the client process (`iperf3 -c 10.0.0.5`) on the host machine (where docker container is running) hits 100% core. `strace -p <iperf3-process>` shows the process is on select() call (meaning it is doing network IO), although reported throughput on the server is 0 kbps. I am not sure if this can be due to closed port issue (in firewall).

**Comment by stevenvanrossem** [comment]
A couple of things here:
* Be aware that the `sonata-fw-vnf` is only a 'dummy' firewall. It is basically a user-space ovs instance with 2 interfaces, and where the OpenFlow entries should make up the firewall rules. To debug, you can check the openflow table of the `ovs1` instance in the `fw-vnf`, check if the expected packets are flowing through: `root@fw:~# ovs-ofctl ovs1 dump-flows`

* Have you changed the interface names of the sonata-fw-vnf? This VNF was configured to work with a certain service descriptor. But you need to take care if you use own interface names as you do in your script [here](https://github.com/sonata-nfv/son-examples/blob/master/vnfs/sonata-fw-vnf-docker/start_fw.sh#L20). 
The startup script of the fw expects the interfaces to be named `fwin` and `fwout`, as you can see [here](https://github.com/sonata-nfv/son-examples/blob/master/vnfs/sonata-fw-vnf-docker/start_fw.sh#L20)

* The ovs instance in the fw-vnf is started in userspace, as configured [here](https://github.com/sonata-nfv/son-examples/blob/master/vnfs/sonata-fw-vnf-docker/start_fw.sh#L24). If you comment this line out, ovs is using its kernel module.  This might solve the issue.

* There is a known issue when using ovs in userspace, it can block tcp traffic:
I have encountered the same problem, there seems to be a bug with offloading tcp segments and checksum calculation. (in my setup, udp traffic was still working).
There is some info here:
https://mail.openvswitch.org/pipermail/ovs-dev/2016-April/313263.html
https://github.com/openvswitch/ovs/blob/master/tests/system-userspace-macros.at#L45
In my case, I could enable tcp traffic again by disabling tcp-offload using ethtool:
you need to disable offload tcp checksum on the interfaces inside the containers that connect to a userspace ovs :
`ethtool -K <interface_name> tx off`
(This solution might have a performance penalty...)

**Comment by knodir** [comment]
Thanks, @stevenvanrossem! It was because of userspace datapath (`datapath_type=netdev`). As you pointed out, I removed that [line](https://github.com/sonata-nfv/son-examples/blob/master/vnfs/sonata-fw-vnf-docker/start_fw.sh#L24) and got high throughput (~25 Gbps). However, `ethtool -K <interface_name> tx off` did not help. So, I do not think it is a problem with offloading.

Although the problem got resolved, I'd like to understand the root cause. I do not think iperf3 packet transfer through userspace consumes an entire CPU core since I get 0 Kbps even with 1M udp traffic (`iperf3 -b 1M`).  Any other explanation why removing `datapath_type=netdev` helped?

**Comment by stevenvanrossem** [comment]
The root cause would require further investigation...
Can you check where the high CPU load is coming from? 
eg. is it the iperf client container that is generating the packets? or is it the ovs-userspace container? or somewhere else? It seems packets are being generated but are not reaching their destination somehow, can you monitor with Wireshark up to what point or interfaces the iperf traffic is flowing?

**Comment by knodir** [comment]
I did a bit more investigation, but not too much. CPU load is coming from iperf client. It is generating very large number of packets. With an eye inspected tcpdump output, I think it is over over 100 packets per second.

Packets are reaching input port of the firewall, but not the output port despite the OVS flow [rule](https://github.com/sonata-nfv/son-examples/blob/master/vnfs/sonata-fw-vnf-docker/start_fw.sh#L48) to forward packets. I did not debug further down OVS.

I'll close this issue for now and come back to it if I get time later. I need to work on other parts of my experiment. Thanks for the help, @stevenvanrossem!
